hal/src/nRF52840/exflash_hal.c

```diff
@@ -35,24 +33,76 @@
 #include "gpio_hal.h"
 #include "system_error.h"
 #include "nrf_system_error.h"
+#include "check.h"
+
+#define PRTCL_QSPI_DEFAULT_CINSTR(opc, len) \
+{                                           \
+    .opcode    = (opc),                     \
+    .length    = (len),                     \
+    .io2_level = true,                      \
+    .io3_level = true,                      \
+    .wipwait   = true,                      \
+    .wren      = true                       \
+}
 
 enum qspi_cmds_t {
-    QSPI_STD_CMD_WRSR     = 0x01,
-    QSPI_STD_CMD_WREN     = 0x06,
-    QSPI_STD_CMD_RSTEN    = 0x66,
-    QSPI_STD_CMD_RST      = 0x99,
-    QSPI_MX25_CMD_ENSO    = 0xb1,
-    QSPI_MX25_CMD_EXSO    = 0xc1,
-    QSPI_MX25_CMD_WRSCUR  = 0x2f,
-    QSPI_MX25_CMD_SLEEP   = 0xB9,
-    QSPI_MX25_CMD_READ_ID = 0xAB,
-    QSPI_STD_CMD_PGMERS_SUSPEND = 0xB0
+    QSPI_STD_CMD_WRSR            = 0x01, // Write Status Register
+    QSPI_STD_CMD_WREN            = 0x06, // Write Enable: Sets WEL bit before every write operation
+    QSPI_STD_CMD_RSTEN           = 0x66, // Enable Reset
+    QSPI_STD_CMD_RST             = 0x99, // Reset
+    QSPI_STD_CMD_SLEEP           = 0xB9, // GD25: "Power Down", MX25: "Deep Power Down"
+    QSPI_STD_CMD_READ_ID         = 0x9F, // Read Identification
+
+    QSPI_MX25_CMD_ENSO           = 0xB1, // Enter Secured OTP
+    QSPI_MX25_CMD_EXSO           = 0xC1, // Exit Secured OTP
+    QSPI_MX25_CMD_RDSCUR         = 0x2B, // Read Security Register
+    QSPI_MX25_CMD_WRSCUR         = 0x2F, // Write Security Register (ie lock OTP)
+    QSPI_MX25_CMD_PGMERS_SUSPEND = 0xB0, // Program/Erase Suspend
+
+    QSPI_GD25_CMD_WRSR2          = 0x31, // Write Status Register 2
+    QSPI_GD25_CMD_WRSR3          = 0x11, // Write Status Register 3
+    QSPI_GD25_CMD_SEC_ERASE      = 0x44, // Security Register Erase (ie Erase OTP)
+    QSPI_GD25_CMD_SEC_READ       = 0x48, // Security Register Read
+    QSPI_GD25_CMD_SEC_PROGRAM    = 0x42, // Security Register Program (ie Write OTP)
+    QSPI_GD25_CMD_PGMERS_SUSPEND = 0x75, // Program/Erase Suspend
+};
+
+enum qspi_type_t {
+    QSPI_TYPE_MX25L3233F,
+    QSPI_TYPE_MX25R6435FZNIL0,
+    QSPI_TYPE_GD25WQ64EQFG,
+    QSPI_TYPE_END
 };
 
-static const size_t MX25_OTP_SECTOR_SIZE = 4096 / 8; /* 4Kb or 512B */
+// For any variations between flash parts, set member vars at runtime
+typedef struct  {
+    int type;                      // qspi_type_t to identify what part is present
+    bool part_detected;            // Have we positively identified the flash part yet
+    uint8_t write_opcode;          // Which opcode to use for quad write
+    uint8_t pgmers_suspend_opcode; // Which opcode to use when interrupting pending program/erase command
+    uint8_t sck_delay;             // tSHSL, tWHSL, and tSHWL in number of 16 MHz periods (62.5ns)
+    uint32_t otp_size;             // Size of OTP area in bytes
+    int (*init)();                 // Function to call once chip is identified. IE enter QSPI mode
```

Comment:
Prefer use typedef.

@YutingYou As discussed, we're going to go with this way:
1. define `hal_qspi_flash_params_t`, which contains flash-specific parameters.
2. create an static `hal_qspi_flash_params_t` array that has all kind of flash parameters initialized
3. define `hal_exflash_t`:
```c++
typedef struct hal_exflash_t {
    hal_qspi_flash_type_t type;
    uintt8_t reserved[3];
    hal_qspi_flash_parames* params;
} hal_exflash_t;
``` 
4. use default QSPI config to read chip ID and initialize the `params` field accordingly followed by initializing QSPI with  new config.